interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodecompilation/BaseNodeCompiler.scala

```diff
@@ -0,0 +1,325 @@
+    val objCases = caseExpressions.map(_.compiledObject).sequence
+
+    val compilationResult = NodeCompilationResult(
+      expressionTypingInfos,
+      None,
+      expressionCompilation.map(_.validationContext).getOrElse(Valid(ctx)),
+      objExpression.product(objCases),
+      expressionCompilation.flatMap(_.expressionType)
+    )
+
+    combineErrors(compilationResult, additionalValidations.combineAll)
+  }
+
+  def compileFields(
+      fields: List[pl.touk.nussknacker.engine.graph.variable.Field],
+      ctx: ValidationContext,
+      outputVar: Option[OutputVar]
+  )(implicit nodeId: NodeId): NodeCompilationResult[List[variable.Field]] = {
+
+    final case class CompiledRecordField(
+        field: pl.touk.nussknacker.engine.graph.variable.Field,
+        index: Int,
+        typedExpression: TypedExpression
+    )
+
+    val compliedRecordFields = fields.zipWithIndex.map { case (field, index) =>
+      objectParametersExpressionCompiler
+        .compile(field.expression, Some(node.recordValueFieldName(index)), ctx, Unknown)
+        .map(CompiledRecordField(field, index, _))
+    }
+
+    val recordValuesCompilationResult = compliedRecordFields.traverse { validatedField =>
+      validatedField.andThen { compiledField =>
+        Valid(
+          ExpressionCompilerAdapter.ExpressionCompilation(
+            compiledField.field.name,
+            Some(compiledField.typedExpression),
+            Valid(variable.Field(compiledField.field.name, compiledField.typedExpression.expression))
+          )
+        )
+      }
+    }
+
+    val typedFieldsValidations = compliedRecordFields.map { validatedField =>
+      validatedField.map { compiledField =>
+        RecordValidator.IndexedField(
+          compiledField.field.name,
+          compiledField.typedExpression,
+          compiledField.index
+        )
+      }
+    }
+    val additionalValidationResult = RecordValidator.validateRecord(
+      typedFieldsValidations.collect { case Valid(typedField) => typedField }
+    )
+
+    val typedObject = recordValuesCompilationResult
+      .map { fieldsComp =>
+        TypedObjectTypingResult(fieldsComp.map(f => (f.fieldName, f.typingResult)).toMap)
+      }
+      .valueOr(_ => Unknown)
+
+    val fieldsTypingInfo = recordValuesCompilationResult
+      .map { compilations =>
+        compilations.flatMap(_.expressionTypingInfo).toMap
+      }
+      .getOrElse(Map.empty)
+
+    val compiledFields = recordValuesCompilationResult.andThen(_.map(_.validated).sequence)
+
+    val compilationResult = NodeCompilationResult(
+      expressionTypingInfo = fieldsTypingInfo,
+      parameters = None,
+      validationContext = outputVar.map(ctx.withVariable(_, typedObject)).getOrElse(Valid(ctx)),
+      compiledObject = compiledFields,
+      expressionType = Some(typedObject)
+    )
+
+    combineErrors(compilationResult, additionalValidationResult)
+  }
+
+  private object ExpressionCompilerAdapter {
+
+    case class ExpressionCompilation[R](
```

Comment:
This name is misleading. Sth is missing here. It is not a "process of compilation itself" but rather some result of compilation? Also, I see some redundancy here. For me, this class looks like a simple wrapper for `ValidatedNel[ProcessCompilationError, TypedExpression]` that additionally holds `fieldName`. If you put this `ValidatedNel` into `newValidated`, you can compute the current validated by `newValidated.map(_.expression)` and `typedExpression` by `newValidated.toOption`.

Reply:
Yes, this is one of the remains from the previous implementation in NodeCompiler. I've reduced it to a simple wrapper and adapted remaining code.